File: 238.product-of-array-except-self.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  var forward = [];
  var backward = [];
  var forw = 1;
  var back = 1;
  for (let i = 0; i < nums.length; i++) {
    forward.push(forw);
    backward.push(back);
    forw *= nums[i];
    back *= nums[nums.length - 1 - i];
  }
  var answer = [];
  for (let i = 0; i < forward.length; i++) {
    answer.push(forward[i] * backward[backward.length - 1 - i]);
  }
  return answer;
};
// @leet end

File: 238.product-of-array-except-self.cpp
// @leet start
#include <vector>
class Solution {
public:
  vector<int> productExceptSelf(vector<int> &nums) {
    std::vector<int> left;
    std::vector<int> right;
    int lefter = 1;
    int righter = 1;
    for (int i = 0; i < nums.size(); i++) {
      left.push_back(lefter);
      right.push_back(righter);
      lefter *= nums[i];
      righter *= nums[nums.size() - i - 1];
    }
    std::vector<int> answer;
    for (int i = 0; i < nums.size(); i++) {
      answer.push_back(left[i] * right[nums.size() - i - 1]);
    }
    return answer;
  }
};
// @leet end
